packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js

```diff
@@ -0,0 +1,307 @@
+ * @property {boolean} canInsertDraggedBlocksAsChild   Whether the dragged block can be a child of this block.
+ */
+
+/**
+ * An object containing details of a drop target.
+ *
+ * @typedef {Object} WPBlockNavigationDropZoneTarget
+ * @property {string}                   blockIndex   The insertion index.
+ * @property {string}                   rootClientId The root client id for the block.
+ * @property {string|undefined}         clientId     The client id for the block.
+ * @property {'top'|'bottom'|'inside'}  dropPosition The position relative to the block that the user is dropping to.
+ *                                                   'inside' refers to nesting as an inner block.
+ */
+
+/**
+ * A react hook that returns data about blocks used for computing where a user
+ * can drop to when dragging and dropping blocks.
+ *
+ * @param {Object}          ref           A React ref of a containing element for block navigation.
+ * @param {WPPoint}         position      The current drag position.
+ * @param {WPDragEventType} dragEventType The drag event type.
+ *
+ * @return {RefObject<WPBlockNavigationDropZoneBlocks>} A React ref containing the blocks data.
+ */
+function useDropTargetBlocksData( ref, position, dragEventType ) {
+	const {
+		getBlockRootClientId,
+		getBlockIndex,
+		getBlockCount,
+		getDraggedBlockClientIds,
+		canInsertBlocks,
+	} = useSelect( ( select ) => {
+		const selectors = select( 'core/block-editor' );
+		return {
+			canInsertBlocks: selectors.canInsertBlocks,
+			getBlockRootClientId: selectors.getBlockRootClientId,
+			getBlockIndex: selectors.getBlockIndex,
+			getBlockCount: selectors.getBlockCount,
+			getDraggedBlockClientIds: selectors.getDraggedBlockClientIds,
+		};
+	}, [] );
+	const blocksData = useRef();
+
+	// Compute data about blocks only when the user
+	// starts dragging, as determined by `hasPosition`.
+	const hasPosition = !! position;
+
+	useEffect( () => {
+		if ( ! ref.current || ! hasPosition ) {
+			return;
+		}
+
+		const isBlockDrag = dragEventType === 'default';
+
+		const draggedBlockClientIds = isBlockDrag
+			? getDraggedBlockClientIds()
+			: undefined;
+
+		const blockElements = Array.from(
+			ref.current.querySelectorAll( '[data-block]' )
+		);
+
+		blocksData.current = blockElements.map( ( blockElement ) => {
+			const clientId = blockElement.dataset.block;
+			const rootClientId = getBlockRootClientId( clientId );
+
+			return {
+				clientId,
+				rootClientId,
+				blockIndex: getBlockIndex( clientId, rootClientId ),
+				element: blockElement,
+				isDraggedBlock: isBlockDrag
+					? draggedBlockClientIds.includes( clientId )
+					: false,
+				innerBlockCount: getBlockCount( clientId ),
+				canInsertDraggedBlocksAsSibling: isBlockDrag
+					? canInsertBlocks( draggedBlockClientIds, rootClientId )
+					: true,
+				canInsertDraggedBlocksAsChild: isBlockDrag
+					? canInsertBlocks( draggedBlockClientIds, clientId )
+					: true,
+			};
+		} );
+	}, [ hasPosition ] );
```

Comment:
I think even with `useEffect`, we still can't re-trigger the effect when ref changed? 🤔

Reply:
Opened #24914 to enable the eslint rules to help us catch these kinds of issues.

Reply:
I am definitely making an assumption that the DOM won't change during a drag/drop based on the fact that the user can't really manipulate the UI in any other way when dragging. 

But at the same time, the dependency list for a hook isn't capable of handling this situation, because part of the subtree of the DOM element captured by ref could mutate, but the reference for the root element may still be the same.

For absolute correctness, a `MutationObserver` would probably be the best option here.

Reply:
IMHO, we can make such assumption, my suggestion is more towards code smell rather than correctness. In case when we did change the DOM in the future. Hence, I don't think we should use `MutationObserver` at this point :)